homeassistant/components/songpal/coordinator.py

```diff
@@ -0,0 +1,264 @@
+                    # We need to inform HA about the state in case we are coming
+                    # back from a disconnected state.
+                    await self.full_refresh()
+                    self.async_set_updated_data(self.data)
+
+            self.hass.loop.create_task(self.device.listen_notifications())
+            _LOGGER.warning(
+                "[%s(%s)] Connection reestablished", self.name, self.device.endpoint
+            )
+
+        self.device.on_notification(VolumeChange, _volume_changed)
+        self.device.on_notification(ContentChange, _source_changed)
+        self.device.on_notification(PowerChange, _power_changed)
+        self.device.on_notification(SettingChange, _setting_changed)
+        self.device.on_notification(ConnectChange, _try_reconnect)
+
+        async def handle_stop(event):
+            await self.device.stop_listen_notifications()
+
+        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, handle_stop)
+
+        self.hass.loop.create_task(self.device.listen_notifications())
+
+    @property
+    def unique_id(self):
+        """Return a unique ID."""
+        return f"{DOMAIN}-{device_unique_id(self.data)}-coordinator"
+
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return the device info."""
+        return device_info(self.device_name, self.data)
+
+    async def async_set_sound_setting(self, name, value):
+        """Change a setting on the device."""
+        _LOGGER.debug("Calling set_sound_setting with %s: %s", name, value)
+        await self.device.set_sound_settings(name, value)
+
+    async def polling_refresh(self, data: dict[str, Any]) -> dict[str, Any]:
+        """Fetch non-pushed updates from the device."""
+        try:
+            data["sound_settings"] = await self.device.get_sound_settings()
+            data["bluetooth_settings"] = await self.device.get_bluetooth_settings()
+            data["misc_settings"] = await self.device.get_misc_settings()
+            data["playback_settings"] = await self.device.get_playback_settings()
+            data["eq"] = await self.device.get_custom_eq()
+            data["play_info"] = await self.device.get_play_info()
+        except SongpalException as ex:
+            _LOGGER.error("Unable to refresh songpal state: %s", ex)
+            self.available = False
+
+        return data
+
+    async def full_refresh(self) -> dict[str, Any]:
+        """Fetch full state from the device."""
+        try:
+            data = self.data
+
+            if "sysinfo" not in data:
+                data["sysinfo"] = await self.device.get_system_info()
+
+            if "interface_info" not in data:
+                data["interface_info"] = await self.device.get_interface_information()
+
+            volumes = await self.device.get_volume_information()
+            if not volumes:
+                _LOGGER.error("Got no volume controls, bailing out")
+                self.available = False
+                return data
```

Comment:
Unsure, but should we really still bail out here if only a single call fails? Or should we have a similar behaviour for the other calls, and maybe even clear the internal state?